agent/proxycfg/manager.go

```diff
@@ -0,0 +1,350 @@
+	// want to maintain a full copy of the state in order to diff and figure out
+	// what changed. Luckily each service has it's own WatchCh so we can figure
+	// out changes and removals with those efficiently.
+	stateCh chan struct{}
+
+	mu       sync.Mutex
+	started  bool
+	proxies  map[string]*state
+	watchers map[string]map[uint64]chan *ConfigSnapshot
+}
+
+// ManagerConfig holds the required external dependencies for a Manager
+// instance. All fields must be set to something valid or the manager will
+// panic. The ManagerConfig is passed by value to NewManager so the passed value
+// can be mutated safely.
+type ManagerConfig struct {
+	// Cache is the agent's cache instance that can be used to retrieve, store and
+	// monitor state for the proxies.
+	Cache *cache.Cache
+	// state is the agent's local state to be watched for new proxy registrations.
+	State *local.State
+	// source describes the current agent's identity, it's used directly for
+	// prepared query discovery but also indirectly as a way to pass current
+	// Datacenter name into other request types that need it. This is sufficient
+	// for now and cleaner than passing the entire RuntimeConfig.
+	Source *structs.QuerySource
+	// logger is the agent's logger to be used for logging logs.
+	Logger *log.Logger
+}
+
+// NewManager constructs a manager from the provided agent cache.
+func NewManager(cfg ManagerConfig) (*Manager, error) {
+	if cfg.Cache == nil || cfg.State == nil || cfg.Source == nil ||
+		cfg.Logger == nil {
+		return nil, errors.New("all ManagerConfig fields must be provided")
+	}
+	m := &Manager{
+		ManagerConfig: cfg,
+		// Single item buffer is enough since there is no data transferred so this
+		// is "level triggering" and we can't miss actual data.
+		stateCh:  make(chan struct{}, 1),
+		proxies:  make(map[string]*state),
+		watchers: make(map[string]map[uint64]chan *ConfigSnapshot),
+	}
+	return m, nil
+}
+
+// Run is the long-running method that handles state syncing. It should be run
+// in it's own goroutine and will continue until a fatal error is hit or Close
+// is called. Run will return an error if it is called more than once, or called
+// after Close.
+func (m *Manager) Run() error {
+	m.mu.Lock()
+	alreadyStarted := m.started
+	m.started = true
+	stateCh := m.stateCh
+	m.mu.Unlock()
+
+	// Protect against multiple Run calls.
+	if alreadyStarted {
+		return ErrStarted
+	}
+
+	// Protect against being run after Close.
+	if stateCh == nil {
+		return ErrStopped
+	}
+
+	// Register for notifications about state changes
+	m.State.Notify(stateCh)
+	defer m.State.StopNotify(stateCh)
+
+	for {
+		m.mu.Lock()
+
+		// Traverse the local state and ensure all proxy services are registered
+		services := m.State.Services()
+		for svcID, svc := range services {
+			if svc.Kind != structs.ServiceKindConnectProxy {
+				continue
+			}
+			// TODO(banks): need to work out when to default some stuff. For example
```

Comment:
TODO now or later?

Reply:
Later. There is a lot we need to do to get little bits of custom config working well...